pageserver/src/remote_storage/storage_sync/upload.rs

```diff
@@ -127,6 +128,12 @@ pub(super) async fn upload_timeline_checkpoint<
                     index_write.add_timeline_entry(sync_id, TimelineIndexEntry::Full(new_timeline));
                 }
             }
+            // TODO (current) Timeline should be available.
+            // We cant just pass Timeline from schedule_checkpoint_upload because checkpoint is defined on a Timeline rather than on Arc<dyn Timeline>.
+            let timeline = tenant_mgr::get_repository_for_tenant(tenant_id)
+                .and_then(|repo| repo.get_local_timeline_load(timeline_id))
+                .expect("timeline should be available because just did a checkpoint");
+            timeline.set_remote_consistent_lsn(new_checkpoint.metadata.disk_consistent_lsn());
```

Comment:
It looks a bit odd that we now have two mechanisms to pass the remote_consistent_lsn:
* the old way via `new_timeline_states` (where we pass on the latest synced timeline_id)
* this way, where we potentially block the executor by locking the timelines

IMO, there has to be one way only to do so, maybe with locking, but not sure if the timeline lock is a good idea: I'd store the data about timeline availability and LSN's somewhere in a separately managed index in memory.

Reply:
Yeah, I tried to pass Arc<dyn TImeline> and atomically update the lsn without locking, but that approach failed with the reason described in the comment, thanks for pointing that out

Reply:
This was changed in a second iteration, so no longer applicable